witness/witness/cmd/witness/internal/witness/witness.go

```diff
@@ -0,0 +1,217 @@
+)
+
+// Opts is the options passed to a witness.
+type Opts struct {
+	DB        *sql.DB
+	PrivKey   string
+	KnownLogs map[string]ct.SignatureVerifier
+}
+
+// Witness consists of a database for storing STHs, a signing key, and a list
+// of logs for which it stores and verifies STHs.
+type Witness struct {
+	db   *sql.DB
+	sk   string
+	Logs map[string]ct.SignatureVerifier
+}
+
+// New creates a new witness, which initially has no logs to follow.
+func New(wo Opts) (*Witness, error) {
+	// Create the sths table if needed.
+	_, err := wo.DB.Exec(`CREATE TABLE IF NOT EXISTS sths (logID BLOB PRIMARY KEY, sth BLOB)`)
+	if err != nil {
+		return nil, err
+	}
+	return &Witness{
+		db:   wo.DB,
+		sk:   wo.PrivKey,
+		Logs: wo.KnownLogs,
+	}, nil
+}
+
+// parse verifies the STH under the appropriate key for logID and returns
+// the parsed STH.
+func (w *Witness) parse(sthRaw []byte, logID string) (*ct.SignedTreeHead, error) {
+	sv, ok := w.Logs[logID]
+	if !ok {
+		return nil, fmt.Errorf("log %q not found", logID)
+	}
+	var sthJSON ct.GetSTHResponse
+	if err := json.Unmarshal(sthRaw, &sthJSON); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal json: %v", err)
+	}
+	sth, err := sthJSON.ToSignedTreeHead()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create STH: %v", err)
+	}
+	if err := sv.VerifySTHSignature(*sth); err != nil {
+		return nil, fmt.Errorf("failed to verify STH signature: %v", err)
+	}
+	return sth, nil
+}
+
+// GetLogs returns a list of all logs the witness is aware of.
+func (w *Witness) GetLogs() ([]string, error) {
+	rows, err := w.db.Query("SELECT logID FROM sths")
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var logs []string
+	for rows.Next() {
+		var logID string
+		err := rows.Scan(&logID)
+		if err != nil {
+			return nil, err
+		}
+		logs = append(logs, logID)
+	}
+	if err := rows.Err(); err != nil {
+		return nil, err
+	}
+	return logs, nil
+}
+
+// GetSTH gets an STH for a given log, which is consistent with all
+// other STHs for the same log signed by this witness.
+func (w *Witness) GetSTH(logID string) ([]byte, error) {
+	sth, err := w.getLatestSTH(w.db.QueryRow, logID)
+	if err != nil {
+		return nil, err
+	}
+	return sth, nil
+}
+
+// Update updates the latest STH if nextRaw is consistent with the current
+// latest one for this log. It returns the latest cosigned STH held by
+// the witness, which is a signed version of nextRaw if the update was applied.
+func (w *Witness) Update(ctx context.Context, logID string, nextRaw []byte, proof [][]byte) ([]byte, error) {
+	// If we don't witness this log then no point in going further.
+	_, ok := w.Logs[logID]
+	if !ok {
+		return nil, status.Errorf(codes.NotFound, "log %q not found", logID)
+	}
+	// Check the signatures on the raw STH and parse it into the STH format.
+	next, err := w.parse(nextRaw, logID)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't parse input STH: %v", err)
+	}
+	// Optimistically sign the new STH now.
```

Comment:
I agree that it could waste some time, but I don't quite see how it's dangerous?  If the concern is a DoS attack then it doesn't seem to increase the vulnerability.

Anyway, the main reason this code is here is to avoid duplicating it in the two places below that we store the STH.  This is also what the generic witness does so maybe if I change it here I should change it there too?

Reply:
Even if it's only currently going to exist in RAM, we will have made a non-repudiable statement that this checkpoint is consistent with previous ones without having verified that it's true... it's relatively "safe" at the moment, but future-us could write a bug which accidentally stores/returns it, so better to just not sign until we know and help future-us :)

If the generic witness does this too, then I agree we should change that too (sorry I missed that if I reviewed that code too!)

Reply:
Okay, just pushed the change.  Let me know if you're happy with it and if so I can raise an analogous PR for the generic witness too!